File: python/trees/serialize.py
class Codec:
    def serialize(self, root):
        result = []

        def dfs(node):
            if not node:
                result.append("None")
                return
            result.append(str(node.val) + ", ")
            dfs(node.left)
            dfs(node.right)
            return

        dfs(root)

        return join(result)

    def deserialize(self, data):
        values = data.split(",")

        def dfs():
            if not values:
                return None
            value = values.pop(0)
            if value == "None":
                return None
            node = TreeNode(int(value))
            node.left = dfs()
            node.right = dfs()
            return node

        return dfs()

File: js/src/trees/serialize.js
const serialize = (root) => {
  const dfs = (node) => {
    if (!node) {
      return [...str, "n"];
    }
    return `${node.val}, ${dfs(node.left)}${dfs(node.right)}`;
  };
  return dfs(root);
};

const deserialize = (data) => {
  const nodes = data.split(",");
  const dfs = () => {
    const nodeStr = nodes.shift();
    if (nodeStr === "n") return null;
    const node = new TreeNode(parseInt(nodeStr, 10));
    node.left = dfs();
    node.right = dfs();
    return node;
  };
  return dfs();
};
